### PR TITLE
Union structural operators

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:

Adds support for a new set of structural operators useful in the development of the tracing app. These operators return matching spans from both the left and right hand side. Previously all structural operators returned right hand side only. They are useful when you want to request something like a failing endpoint AND all descendant failing spans in one query:

```
{ span.http.url = "/path/of/api" && status = error } &>> { status = error }
```

~Also, rewrites the existing operators to be more performant. Benchmarks incoming.~

Early benchmarks were showing very nice performance gains using the new implementations. However, I have since found this is heavily dependent on trace structure. I have reverted non-union operations to their old form as they seemed to perform more consistently against a variety of trace structures.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by mdisibio** [comment]
Ran everything locally and the new operators are working well.  Now will dig into the loop code further, but wanted to leave an update in the meantime.

File: docs/sources/tempo/traceql/_index.md [hunk 1]
```diff
@@ -262,6 +262,23 @@ For example, to find a trace where a specific HTTP API interacted with a specifi
 { span.http.url = "/path/of/api" } >> { span.db.name = "db-shard-001" }
 ```
 
+### Union Structural
```
Comment (knylander-grafana) [inline docs/sources/tempo/traceql/_index.md:265]: Changed to sentence case. 
```suggestion
### Union structural
```

File: docs/sources/tempo/traceql/_index.md [hunk 2]
```diff
@@ -262,6 +262,23 @@ For example, to find a trace where a specific HTTP API interacted with a specifi
...
+
+These spanset operators look at the structure of a trace and the relationship between the spans. These operators are unique in that they
+return spans that match on both side of the operator. 
+
+- `{condA} &>> {condB}` - The descendant operator (`>>`) looks for spans matching `{condB}` that are descendants of a span matching `{condA}`
+- `{condA} &<< {condB}` - The ancestor operator (`<<`) looks for spans matching `{condB}` that are ancestor of a span matching `{condA}`
+- `{condA} &> {condB}` - The child operator (`>`) looks for spans matching `{condB}` that are direct child spans of a parent matching `{condA}`
+- `{condA} &< {condB}` - The parent operator (`<`) looks for spans matching `{condB}` that are direct parent spans of a child matching `{condA}`
```
Comment (knylander-grafana) [inline docs/sources/tempo/traceql/_index.md:273]: Add missing periods. 
```suggestion
- `{condA} &>> {condB}` - The descendant operator (`>>`) looks for spans matching `{condB}` that are descendants of a span matching `{condA}`.
- `{condA} &<< {condB}` - The ancestor operator (`<<`) looks for spans matching `{condB}` that are ancestor of a span matching `{condA}`.
- `{condA} &> {condB}` - The child operator (`>`) looks for spans matching `{condB}` that are direct child spans of a parent matching `{condA}`.
- `{condA} &< {condB}` - The parent operator (`<`) looks for spans matching `{condB}` that are direct parent spans of a child matching `{condA}`.
```

**Review by knylander-grafana — COMMENTED** [review]
Thank you for updating the doc! Approving doc updates but not the full PR.

**Review by mdisibio — APPROVED** [review]

**Pull request merged**